py_ecc/BaseCurve.py

```diff
@@ -0,0 +1,291 @@
+        """
+        Check that a point is on the curve
+        """
+        if self.is_inf(pt):
+            return True
+        x, y = pt
+        return y**2 == x**3 + b
+
+    def double(self, pt: Point2D[Field]) -> Point2D[Field]:
+        """
+        Elliptic Curve Doubling (P+P).
+        """
+        x, y = pt
+        m = (3 * x**2) / (2 * y)
+        newx = m**2 - 2 * x
+        newy = -m * newx + m * x - y
+        return (newx, newy)
+
+    def add(self,
+            p1: Point2D[Field],
+            p2: Point2D[Field]) -> Point2D[Field]:
+        """
+        Elliptic curve addition.
+        """
+        if p1 is None or p2 is None:
+            return p1 if p2 is None else p2
+        x1, y1 = p1
+        x2, y2 = p2
+        if x2 == x1 and y2 == y1:
+            return self.double(p1)
+        elif x2 == x1:
+            return None
+        else:
+            m = (y2 - y1) / (x2 - x1)
+        newx = m**2 - x1 - x2
+        newy = -m * newx + m * x1 - y1
+        assert newy == (-m * newx + m * x2 - y2)
+        return (newx, newy)
+
+    def multiply(self, pt: Point2D[Field], n: int) -> Point2D[Field]:
+        """
+        Elliptic curve point multiplication.
+        """
+        if n == 0:
+            return None
+        elif n == 1:
+            return pt
+        elif not n % 2:
+            # print(n//2)
+            return self.multiply(self.double(pt), n // 2)
+        else:
+            # print(n//2)
+            return self.add(self.multiply(self.double(pt), n // 2), pt)
+
+    def eq(self, p1: GeneralPoint[Field], p2: GeneralPoint[Field]) -> bool:
+        """
+        Check if 2 points are equal.
+        """
+        return p1 == p2
+
+    def neg(self, pt: Point2D[Field]) -> Point2D[Field]:
+        """
+        Gives the reflection of point wrt x-axis (P => -P).
+        """
+        if pt is None:
+            return None
+        x, y = pt
+        return (x, -y)
+
+    @abstractmethod
+    def twist(self, pt: Point2D[FQP]) -> Point2D[FQP]:
+        """
+        'Twist' a point in E(FQ2) into a point in E(FQ12)
+        """
+        raise NotImplementedError("Must be implemented by subclasses")
+
+
+class BaseOptimizedCurve:
+    # Name of the curve can be "bn128" or "bls12_381"
+    curve_name = None
+    curve_order = None
+    # Curve is y**2 = x**3 + b
+    b = None
+    # Twisted curve over FQ**2
+    b2 = None
+    # Extension curve over FQ**12; same b value as over FQ
+    b12 = None
+    # Generator for curve over FQ
+    G1 = None
+    # Generator for twisted curve over FQ2
+    G2 = None
+    # Generator for curve over FQ12
+    G12 = None
+    # Point at infinity over FQ
+    Z1 = None
+    # Point at infinity for twisted curve over FQ2
+    Z2 = None
+
+    def __init__(self, curve_properties, curve_name):
+        self.curve_name = curve_name
+        self.curve_order = curve_properties[curve_name]["curve_order"]
+        self.b = curve_properties[curve_name]["b"]
```

Comment:
All of these look like they'd be better done as properties of the class itself rather than needing to instantiate the class to get access to them.  Then we can define a subclass of the `BaseCurve` class for each curve we support so that we don't have to manage instantiating each one?